### PR TITLE
support multiple disks in installer

### PR DESCRIPTION
**Pull request by giggsoff**
Support for multiple disks install for persist dataset in with `eve_persist_disk=XXX` option in ipxe/grub. If we defined 2 disks we do mirror, 3 - raidz1, 4 and larger raidz2. If we defined one disk with comma or just comma we will use zfs as well for just one disk/partition. It supports allocation one of dev in dataset to be located on the same block device as EVE boot/rootfs

Signed-off-by: Petr Fedchenkov <giggsoff@gmail.com>

### CODE REVIEW

File: pkg/eve/runme.sh [hunk 1]
```diff
@@ -59,6 +60,8 @@ Passing -v <local folder>:/out makes sure the file created is given most appropr
 
 -f fmt selects a packaging format: raw (default), qcow2, parallels, vdi and gcp are all valid options.
 
+-p disks selects disks to use for persist separated by comma (e.g. sda,sdb,sdc). May be used for installers.
```
Comment (eriknordmark) [inline pkg/eve/runme.sh:63]: Why do we need this? We can already set eve_persist_disk and with these changes that can take a comma-separated list.

File: pkg/mkimage-raw-efi/install [hunk 2]
```diff
@@ -170,24 +171,57 @@ done
...
+if [ "$DISKS_TO_MERGE" != "" ]; then
+  if [ "$DISK_WITH_P3" != "" ]; then
+    P3_ID="$(find_part P3 "$DISK_WITH_P3")"
+    [ -z "$P3_ID" ] && bail "Installation failed. Cannot found P3. Entering shell..."
+    DISKS_TO_MERGE="/dev/$P3_ID $DISKS_TO_MERGE"
+  fi
+  [ -e /root/sys ] || mkdir /root/sys && mount -t sysfs sysfs /root/sys
+  [ -e /root/proc ] || mkdir /root/proc && mount -t proc proc /root/proc
+  [ -e /root/dev ] || mkdir /root/dev && mount -t devtmpfs -o size=10m,nr_inodes=248418,mode=755,nosuid,noexec,relatime devtmpfs /root/dev
+  CMD="chroot /root zpool create -f -m none -o feature@encryption=enabled -O overlay=on persist $DISKS_TO_MERGE"
```
Comment (eriknordmark) [inline pkg/mkimage-raw-efi/install:216]: This is just striping across the disks, right? I don't think that is useful. raidz2 would be useful, and raidz if we have too few disks for raidz2. But we would also potentially want to figure out if we have a mixture of SSDs and rotating disks, the we potentially want to place the zil on the ssd(s).

So it isn't clear to me what use case you have in mind for this.

File: pkg/eve/runme.sh (hunk 1 above)
Comment (giggsoff) [reply pkg/eve/runme.sh:63]: With this change we can set it when we create installer (with docker run lfedge/eve…)

File: pkg/mkimage-raw-efi/install (hunk 2 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:216]: Good point. My variant is the simplest option. 
Regarding to the raid I can think about options to add.

Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:217]: What do you think, if I will allow to write something like `eve_persist_disk=raidz1,sda,sdb,sdc,log,mirror,sdd,sde,cache,sdf`?

Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:225]: Now it support complex notations

File: pkg/eve/runme.sh (hunk 1 above)
Comment (eriknordmark) [reply pkg/eve/runme.sh]: But we can accomplish the same by putting it in grub.cfg in either the mounted config partition or in conf/grub.cfg and make.

If we want to be able to do this as part of the docker run it would make sense to add support for adding all of the options (with their existing names) which go into grub.cfg; that way I can also set the eve disk etc, and using the same names as opposed to unique option names for a subset.

File: pkg/mkimage-raw-efi/install (hunk 2 above)
Comment (eriknordmark) [reply pkg/mkimage-raw-efi/install:225]: I'd actually prefer if we figure that out based on the number of disks listed and not have our users have to be experts on how to setup the ZFS pool.

File: pkg/eve/runme.sh (hunk 1 above)
Comment (eriknordmark) [reply pkg/eve/runme.sh]: Instead of -p sda,sdb can we do -o eve_persist_disk=sda,sdb?
That way we can use -o for different grub options and also use it to specify the root disk.

File: pkg/mkimage-raw-efi/install (hunk 2 above)
Comment (eriknordmark) [reply pkg/mkimage-raw-efi/install:225]: So if 2 we do mirror, 3 raidz1, 4 and larger raidz2.
Then we can evolve the default selection over time.

File: pkg/eve/runme.sh (hunk 1 above)
Comment (giggsoff) [reply pkg/eve/runme.sh]: Done

File: pkg/mkimage-raw-efi/install (hunk 2 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install]: Modified

File: pkg/eve/runme.sh [hunk 3]
```diff
@@ -59,6 +60,8 @@ Passing -v <local folder>:/out makes sure the file created is given most appropr
 
 -f fmt selects a packaging format: raw (default), qcow2, parallels, vdi and gcp are all valid options.
 
+-o option provides an option to installer cmdline (e.g. -o eve_persist_disk=sda,sdb -o eve_install_disk=sda). May be used for installers.
```
Comment (eriknordmark) [inline pkg/eve/runme.sh:63]: Would this example work with sda in two places? Or should it say sdb.sdc for the persist disk?

**Review by eriknordmark — APPROVED** [review]
LGTM

File: pkg/eve/runme.sh [hunk 4]
```diff
@@ -48,7 +49,7 @@ dump() {
 
 do_help() {
 cat <<__EOT__
-Usage: docker run lfedge/eve [-f fmt] version|rootfs|live|installer_raw|installer_iso|installer_net
+Usage: docker run lfedge/eve [-f fmt] [-o option] version|rootfs|live|installer_raw|installer_iso|installer_net
```
Comment (rvs) [inline pkg/eve/runme.sh:52]: Honestly -- I'm NOT at all a big fan of this being an option to the lfedge/eve container @giggsoff -- personally I think it sort of breaks the customization model of where the images are expected to be more or less generic and all customizations are done post producing the image.

The option also breaks orthogonality (since it isn't really applicable to non-installer images).

Therefore two questions:
   1. can we do this differently?
   2. if there's a good 'no' answer to the above -- I honestly think we should NOT document it and keep thinking about a good 'yes' answer to #1

File: pkg/mkimage-raw-efi/install [hunk 5]
```diff
@@ -70,7 +70,8 @@ root_dev() {
 find_part() {
    PARTS=$(lsblk -anl -o "NAME,PARTLABEL" 2>/dev/null | sed -ne "/$1"'$/s#'"$1##p")
    for p in $PARTS ; do
-      [ -f "/sys/block/$2/$p/dev" ] && echo "$p" && exit 0
+      PART=$(echo "$p"|tr -d '[:space:]')
```
Comment (rvs) [inline pkg/mkimage-raw-efi/install:73]: why is this needed here @giggsoff ?

File: pkg/mkimage-raw-efi/install [hunk 6]
```diff
@@ -170,24 +171,82 @@ done
...
 if [ "$INSTALL_DEV" != "$INSTALL_PERSIST" ]; then
-   # apparently sgdisk -Z doesn't clear MBR and keeps complaining
-   PDEV="/dev/$INSTALL_PERSIST"
-   dd if=/dev/zero of="$PDEV" bs=512 count=1 conv=notrunc
-   sgdisk -Z --clear "$PDEV" 2>/dev/null || :
-   sgdisk --new 1:2048:0 --typecode=1:5f24425a-2dfa-11e8-a270-7b663faccc2c --change-name=1:P3 "$PDEV"
-   sgdisk -v "$PDEV"
-   # force make-raw to skip persist
-   MAKE_RAW_PARTS="efi imga imgb conf"
+   if echo "$INSTALL_PERSIST"| grep -q ","; then
```
Comment (rvs) [inline pkg/mkimage-raw-efi/install:179]: So what if I want zfs on a single P3 disk? I won't have a comma in my spec will I?

File: pkg/mkimage-raw-efi/install [hunk 7]
```diff
@@ -170,24 +171,82 @@ done
...
+      POOL_CREATION_COMMAND="mirror $POOL_CREATION_COMMAND"
+      ;;
+    3)
+      POOL_CREATION_COMMAND="raidz1 $POOL_CREATION_COMMAND"
+      ;;
+    *)
+      POOL_CREATION_COMMAND="raidz2 $POOL_CREATION_COMMAND"
+      ;;
+  esac
+  [ -e /root/sys ] || mkdir /root/sys && mount -t sysfs sysfs /root/sys
```
Comment (rvs) [inline pkg/mkimage-raw-efi/install:237]: Ok, all of this needs to be either put into a function -- or better yet -- we need eve(1) exec-like script in the install container

File: pkg/mkimage-raw-efi/install [hunk 8]
```diff
@@ -208,14 +267,23 @@ if mount_part INVENTORY "$(root_dev)" -t vfat -o iocharset=iso8859-1; then
...
-grep -q eve_pause_after_install /proc/cmdline && pause "before shutting the node down"
+grep -q eve_pause_after_install /proc/cmdline && pause "shutting the node down"
 
 # lets hope this is enough to flush the caches
 sync; sleep 5; sync
 for p in CONFIG INVENTORY P3; do
    umount "/run/$p" 2>/dev/null
 done
 
+if [ "$POOL_CREATION_COMMAND" != "" ] && [ $DISKS_TO_MERGE_COUNT -gt 1 ]; then
```
Comment (rvs) [inline pkg/mkimage-raw-efi/install:278]: Why is this delayed to this point?

**Review by rvs — CHANGES_REQUESTED** [review]
Hey @giggsoff -- while it may sounds silly -- I really don't want `-o` to be user visible for now (see my comments). That part needs to be taken care of.

There's also a few things when it comes to cleaning up our shell scriptting.

File: pkg/eve/runme.sh (hunk 4 above)
Comment (giggsoff) [reply pkg/eve/runme.sh:52]: The same things may be done by manually modifying of grub.cfg on config partition or ipxe.cfg. So, it was added here for just convenience.
I do not fully understand the comment (the second option), do you propose to remove the comment or completely remove this feature from this PR?

File: pkg/mkimage-raw-efi/install (hunk 6 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:179]: I keep the old variant of deployment in this case, so, if EVE will see zfs-(kvm|xen|acrn) in the eve-release, storage-init will allocate zfs on the partition/disk.

File: pkg/mkimage-raw-efi/install (hunk 5 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:73]: Because of formatting inside `lsblk`, see result from my PC (spaces between name and partlabel):
```
sda       
sda1      Microsoft reserved partition
sda2      Basic data partition
sda3      
sda5      
nvme0n1   
nvme0n1p1 Basic data partition
nvme0n1p2 EFI system partition
nvme0n1p3 Microsoft reserved partition
nvme0n1p4 Basic data partition
nvme0n1p5 
```

File: pkg/mkimage-raw-efi/install (hunk 8 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:278]: Because we want to have `collect_black_box` before unmount.

File: pkg/eve/installer/grub.cfg [hunk 9]
```diff
@@ -30,21 +30,26 @@ function installer_submenus {
...
+      set_global dom0_cmdline "$dom0_cmdline eve_persist_disk=$eve_persist_disk"
+   }
 }
 
 function set_global {
    set $1="$2"
    export $1
 }
 
+dom0_additional_args=
```
[the remark this answers is not in the document]
Comment (giggsoff) [reply pkg/eve/installer/grub.cfg:44]: To override from `-o option`. But if we will remove such behavior, it will gone (please see next comment).

File: pkg/mkimage-raw-efi/install (hunk 7 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:237]: Will move it to a function as a first variant.

File: pkg/eve/runme.sh (hunk 3 above)
Comment (giggsoff) [reply pkg/eve/runme.sh:63]: Yes, it will work. In this case we will have /dev/sda9 in the pool along with other disks.

File: pkg/eve/runme.sh (hunk 4 above)
Comment (rvs) [reply pkg/eve/runme.sh:52]: I'd rather completely remove the feature for now -- but if it is somehow going to make your workflow super difficult we can settle on removing the user-visible indication of the feature.

File: pkg/mkimage-raw-efi/install (hunk 5 above)
Comment (rvs) [reply pkg/mkimage-raw-efi/install:73]: Sorry for being dense -- but the first column doesn't seem to ever contain spaces and ```$p``` is iterating over the first column, correct?

File: pkg/mkimage-raw-efi/install (hunk 6 above)
Comment (rvs) [reply pkg/mkimage-raw-efi/install:179]: Can we somehow unify the single disk behaviour as well? I mean -- I'd be even fine with ```...=sda,``` so that the presence of the coma (even for a single disk) would trigger this

File: pkg/mkimage-raw-efi/install (hunk 8 above)
Comment (rvs) [reply pkg/mkimage-raw-efi/install:278]: got it!

File: pkg/mkimage-raw-efi/install (hunk 7 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install]: Moved

File: pkg/mkimage-raw-efi/install (hunk 6 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install:199]: Added support for this. You can just place comma to use partition on the same disk as an eve_install_disk.

File: pkg/eve/installer/grub.cfg (hunk 9 above)
[the remark this answers is not in the document]
Comment (giggsoff) [reply pkg/eve/installer/grub.cfg]: Removed for now

File: pkg/eve/runme.sh (hunk 4 above)
Comment (giggsoff) [reply pkg/eve/runme.sh]: I removed them for now in favor of modification of grub/ipxe manually

File: pkg/mkimage-raw-efi/install (hunk 5 above)
Comment (giggsoff) [reply pkg/mkimage-raw-efi/install]: The problem here in spaces added to $p (something like "sda1      " and "nvme0n1p3 " from my previous sample. Please see new version in last commit, I put changes in top sed.

**Review by rvs — APPROVED** [review]
LGTM. Let it test! Thanks for taking care of all the feedback @giggsoff
